doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,416 @@
+until the current wall time catches up to the global timestamp. As an optimization we can implement a form of group
+commit, where all writes within the same millisecond commit together at the same timestamp.
+
+**Pros:**
+
+- Prevents reads from being blocked.
+
+**Cons:**
+
+- Limits write throughput (1 write per millisecond without group commit) or limits write latency (1 millisecond per
+  write with group commit).
+
+##### Hybrid Logical Timestamps
+
+Note: There might be an actual name for this concept, but I chose the name above because it’s similar to Hybrid Logical
+Clocks.
+
+Currently, the timestamps used to in Materialize for `upper`, `since`, and the global timestamp are 64 bit unsigned
+integers that are interpreted as an epoch in milliseconds. Instead we can change the representation of these timestamps
+so that the X most significant bits are used for the epoch and the Y least significant bits are used for a counter.
+Every 1 millisecond the X bits are increased by 1 and the Y bytes are all reset to 0. Every write increments the Y bits
+by 1. This would allow us to have 2^Y writes per millisecond, without blocking any queries. Any write in
+excess of 2^Y per millisecond must be blocked until the next millisecond.
+
+Note: Increasing the size of the timestamp type would give more flexibility (for example to 128 bits).
+
+**Pros:**
+
+- With a sufficiently high Y, no queries would be blocked.
+
+**Cons:**
+
+- Complicated to implement and likely has a large blast radius.
+- If X is low enough we will overflow our timestamps at some point in the future. This can be mitigated with a high
+  enough X.
+
+##### Write Ahead Log (WAL)
+
+This approach involves the coordinator buffering all writes in a durable write ahead log. All writes that occur in the
+same millisecond can be assigned the same timestamp, as long as there are no reads in-between them. Then at the end of
+the millisecond, or after X milliseconds, all writes are sent to STORAGE in a batch. If Materialize crashes/restarts
+before sending the writes to STORAGE, then it can just read the unsent write from the durable logs after starting back
+up. This prevents us from having to increment the global timestamp on every write. However, this does not solve the
+problem for read/write cycles.
+
+**Pros:**
+
+- No queries are blocked from consecutive writes.
+- A WAL would be useful for many other reasons not in scope for this document. It’s likely that no matter what solution
+  we pick, we’d still want to implement a WAL.
+
+**Cons:**
+
+- Only solves half the problem.
+
+##### Timely Fast Forward
+
+Timestamps in Materialize are meant to closely track the wall clock, but are not guaranteed to be exactly equal to the
+wall clock. This approach would involve forcibly fast-forwarding the `upper`s of all objects. Whenever a write would
+push the global timestamp above the wall clock by some threshold, all `upper`s are jumped to the global timestamp’s new
```

Comment:
I'm not sure this is in the cards at the moment. Sources are in charge of how they generate timestamps, and this change would add coord to that mix, which we are very unlikely to do now.

Reply:
I thought this one would probably be a longshot.